alerting/alerter-war/src/main/java/org/hawkular/metrics/alerter/groups/GroupTriggerManager.java

```diff
@@ -0,0 +1,597 @@
+                } else {
+                    try {
+                        // start the job.
+                        log.warnf("Adding runner for %s with job period %d seconds", mgt, JOB_PERIOD);
+
+                        MGTRunner runner = new MGTRunner(metrics, definitions, mgt);
+                        mgtFutures.put(
+                                mgt,
+                                mgtExecutor.scheduleAtFixedRate(runner, 0L, JOB_PERIOD, TimeUnit.SECONDS));
+                    } catch (Exception e) {
+                        log.errorf("Failed to schedule metrics group trigger %s: %s ", mgt, e.getMessage());
+                    }
+                }
+            }
+
+            // cancel any job for an mgt not in the current set
+            Set<Trigger> doomedJobs = mgtFutures.keySet().stream()
+                    .filter(k -> !mgts.contains(k))
+                    .collect(Collectors.toSet());
+            doomedJobs.stream().forEach(k -> {
+                log.warnf("Canceling doomed job for %s", k);
+                cancel(k);
+            });
+
+        } catch (Exception e) {
+            log.error("Failed to refresh Metrics Group Triggers.", e);
+        }
+    }
+
+    static class MGTRunner implements Runnable {
+        private final Logger log = Logger.getLogger(GroupTriggerManager.MGTRunner.class);
+
+        private MetricsService metricsService;
+        private DefinitionsService definitionsService;
+        private Trigger metricsGroupTrigger;
+        private Integer dataIdMapHash;
+
+        public MGTRunner(MetricsService metrics, DefinitionsService definitionsService, Trigger metricsGroupTrigger) {
+            super();
+            this.metricsService = metrics;
+            this.definitionsService = definitionsService;
+            this.metricsGroupTrigger = metricsGroupTrigger;
+        }
+
+        @Override
+        public void run() {
+            log.warnf("Running job for MGT %s", metricsGroupTrigger);
+            try {
+                // fetch the trigger to make sure it still exists and is tagged appropriately
+                Trigger mgt = definitionsService.getTrigger(metricsGroupTrigger.getTenantId(),
+                        metricsGroupTrigger.getId());
+
+                if (null == mgt) {
+                    log.warnf("Metrics Group Trigger not found. Skipping %s", mgt);
+                    return;
+                }
+                if (!mgt.isGroup()) {
+                    log.warnf("Metrics Group Trigger is not a group trigger, skipping %s", mgt);
+                    return;
+                }
+                Map<String, String> tags = mgt.getTags();
+                if (isEmpty(tags.getOrDefault(TAG_DATA_IDS, ""))) {
+                    log.warnf("Required Metrics Group Trigger tag [%s] missing or null, skipping %s",
+                            TAG_DATA_IDS, mgt);
+                    return;
+                }
+                if (isEmpty(tags.getOrDefault(TAG_SOURCE_BY, ""))) {
```

Comment:
Good question.  SourceBy is required, and does not default to "*", because I want to make sure that the client explicitly sets "*" when desired.  The reason is to prevent an undesired generation of a large member trigger set.  The wildcard SourceBy basically means use all combinations of the metrics resulting from the dataId tag queries.   For single-dataId-conditions it is likely that a user would use SourceBy="*", but this is just a guard against a potentially bad mistake, particularly in the CompareCondition scenario.